p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,302 @@
+package holepunch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/jpillora/backoff"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	holepunch_pb "github.com/libp2p/go-libp2p/p2p/protocol/holepunch/pb"
+	"github.com/libp2p/go-libp2p/p2p/protocol/identify"
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// TODO Should we have options for these ?
+const (
+	protocol         = "/libp2p/holepunch/1.0.0"
+	maxMsgSize       = 4 * 1024 // 4K
+	holePunchTimeout = 2 * time.Minute
+	dialTimeout      = 60 * time.Second
+	maxRetries       = 4
+)
+
+var (
+	log = logging.Logger("p2p/holepunch")
+)
+
+// TODO Find a better name for this protocol.
+// HolePunchService is used to make direct connections with a peer via hole-punching.
+type HolePunchService struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	ids  *identify.IDService
+	host host.Host
+
+	// ensure we shutdown ONLY once
+	closeSync sync.Once
+	refCount  sync.WaitGroup
+}
+
+// NewHolePunchService creates a new service that can be used for hole punching
+func NewHolePunchService(h host.Host, ids *identify.IDService) (*HolePunchService, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	hs := &HolePunchService{ctx: ctx, ctxCancel: cancel, host: h, ids: ids}
+
+	h.SetStreamHandler(protocol, hs.handleNewStream)
+	h.Network().Notify((*netNotifiee)(hs))
+	return hs, nil
+}
+
+// Close closes the Hole Punch Service.
+func (hs *HolePunchService) Close() error {
+	hs.closeSync.Do(func() {
+		hs.ctxCancel()
+		hs.refCount.Wait()
+	})
+
+	return nil
+}
+
+// attempts to make a direct connection with the remote peer of `relayConn` by co-ordinating a hole punch over
+// the given relay connection `relayConn`.
+func (hs *HolePunchService) holePunch(relayConn network.Conn) {
+	rp := relayConn.RemotePeer()
+
+	// short-circuit hole punching if a direct dial works.
+	// attempt a direct connection ONLY if we have a public address for the remote peer
+	for _, a := range hs.host.Peerstore().Addrs(rp) {
+		if manet.IsPublicAddr(a) && !isRelayAddress(a) {
+			forceDirectConnCtx := network.WithForceDirectDial(hs.ctx, "hole-punching")
+			dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+			defer cancel()
+			if err := hs.host.Connect(dialCtx, peer.AddrInfo{ID: rp}); err == nil {
+				log.Debugf("direct connection to peer %s successful, no need for a hole punch", rp.Pretty())
+				return
+			}
```

Comment:
we should log the error if we fail; also, the dialTimeout needs to be dialed way down.

Reply:
I don't think logging the error is a good idea here. Remember, we also dial the other peer if either of us is behind an `Unknown` NAT.  If it turns out that either of us is infact behind a Symmetric NAT, this will cause unnecessary noise in the logs.

Reply:
ok, fair enough.